pinot-controller/src/main/java/org/apache/pinot/controller/LeadControllerManager.java

```diff
@@ -0,0 +1,138 @@
+  }
+
+  /**
+   * Marks the cached indices invalid and unregisters {@link PinotHelixResourceManager}.
+   */
+  public void stop() {
+    _partitionIndexCache.clear();
+    _pinotHelixResourceManager = null;
+  }
+
+  /**
+   * Checks whether the current controller is the leader for the given table. Return true if current controller is the leader for this table.
+   * Otherwise check whether the current controller is helix leader if the resource is disabled.
+   * @param tableName table name with/without table type.
+   */
+  public boolean isLeaderForTable(String tableName) {
+    if (isLeadControllerResourceEnabled()) {
+      String rawTableName = TableNameBuilder.extractRawTableName(tableName);
+      int partitionIndex = LeadControllerUtils.getPartitionIdForTable(rawTableName, NUMBER_OF_PARTITIONS_IN_LEAD_CONTROLLER_RESOURCE);
+      return _partitionIndexCache.contains(partitionIndex);
+    } else {
+      // Checks if it's Helix leader if lead controller resource is disabled.
+      return isHelixLeader();
+    }
+  }
+
+  /**
+   * Given a partition name, marks current controller as lead controller for this partition by caching the partition index to current controller.
+   * @param partitionName partition name in lead controller resource, e.g. leadControllerResource_0.
+   */
+  public synchronized void addPartitionLeader(String partitionName) {
+    LOGGER.info("Add Partition: {} to LeadControllerManager", partitionName);
+    int partitionIndex = LeadControllerUtils.extractPartitionIndex(partitionName);
+    _partitionIndexCache.add(partitionIndex);
+  }
+
+  /**
+   * Given a partition name, removes current controller as lead controller for this partition by removing the partition index from current controller.
+   * @param partitionName partition name in lead controller resource, e.g. leadControllerResource_0.
+   */
+  public synchronized void removePartitionLeader(String partitionName) {
+    LOGGER.info("Remove Partition: {} from LeadControllerManager", partitionName);
+    int partitionIndex = LeadControllerUtils.extractPartitionIndex(partitionName);
+    _partitionIndexCache.remove(partitionIndex);
+  }
+
+  /**
+   * Checks if the current controller host is Helix cluster leader.
+   */
+  private boolean isHelixLeader() {
+    return _amIHelixLeader;
+  }
+
+  public boolean isLeadControllerResourceEnabled() {
+    return _isLeadControllerResourceEnabled;
+  }
+
+  /**
+   * Callback on changes in the controller. Should be registered to the controller callback. This callback is not needed when the resource is enabled.
+   * However, the resource can be disabled sometime while the cluster is in operation, so we keep it here. Plus, it does not add much overhead.
+   * At some point in future when we stop supporting the disabled resource, we will remove this line altogether and the logic that goes with it.
+   */
+  void onHelixControllerChange() {
+    if (_pinotHelixResourceManager.isHelixLeader()) {
```

Comment:
Even if we introduce a `_isShuttingDown` flag, it could still be possible that `_isShuttingDown` is false and then another thread immediately set it to be true before processing Helix controller change. Added synchronized on the methods.